app/actors/sufia/interpret_visibility_actor.rb

```diff
@@ -0,0 +1,107 @@
+      def validate_release_type(template)
+        # It's valid as long as embargo is not specified when a template requires no release delays
+        return true unless @intention.wants_embargo? && template.present? && template.release_no_delay?
+
+        curation_concern.errors.add(:visibility, 'Visibility specified does not match permission template "no release delay" requirement for selected AdminSet.')
+        false
+      end
+
+      # Validate visibility complies with AdminSet template requirements
+      def validate_visibility(attributes, template)
+        # NOTE: For embargo/lease, attributes[:visibility] will be nil (see sanitize_params), so visibility will be validated as part of embargo/lease
+        return true unless template.present? && attributes[:visibility].present?
+
+        # Validate against template's visibility requirements
+        return true if template.valid_visibility?(attributes[:visibility])
+
+        curation_concern.errors.add(:visibility, 'Visibility specified does not match permission template visibility requirement for selected AdminSet.')
+        false
+      end
+
+      # When specified, validate embargo is a future date that complies with AdminSet template requirements (if any)
+      def validate_embargo(attributes, template)
+        return true unless @intention.wants_embargo?
+
+        embargo_date = parse_date(attributes[:embargo_release_date])
+
+        # When embargo required, date must be in future AND matches any template requirements
+        return true if valid_future_date?(embargo_date) && valid_template_embargo_date?(embargo_date, template) && valid_template_visibility_after_embargo?(attributes, template)
+
+        curation_concern.errors.add(:visibility, 'When setting visibility to "embargo" you must also specify embargo release date.') unless embargo_date.present?
+        false
+      end
+
+      # Validate a date string is both a date AND in the future
+      def valid_future_date?(date)
+        return true if date.present? && date.future?
+
+        curation_concern.errors.add(:embargo_release_date, "Must be a future date.")
+        false
+      end
+
+      # Validate an embargo date against permission template restrictions
+      def valid_template_embargo_date?(date, template)
+        return true unless template.present?
+
+        # Validate against template's release_date requirements
+        return true if template.valid_release_date?(date)
+
+        curation_concern.errors.add(:embargo_release_date, "Release date specified does not match permission template release requirements for selected AdminSet.")
+        false
+      end
+
+      # Validate the post-embargo visibility against permission template requirements (if any)
+      def valid_template_visibility_after_embargo?(attributes, template)
+        return true unless template.present? && attributes[:visibility_after_embargo].present?
```

Comment:
To appease the Code Climate monster (https://codeclimate.com/github/projecthydra/sufia/pull/2966) you could adjust accordingly:

```ruby
def validate_visibility(attributes, template)
  return true unless template_and_attribute_are_present?(: visibility_after_embargo)
  ...
end
```